---
title: "geneset_analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(dplyr)
library(readr)
library(knitr)
library(ggplot2)
library(RColorBrewer)
library(ggsci)
library(cowplot)
library(biomaRt)
library(scales)
library(matrixStats)
library(gtools)
library(parallel)
library(tidytext)
library(msigdbr)
library(purrr)
library(ggvenn)

rm(list=ls())

opts_knit$set(root.dir ='/ru-auth/local/home/lzhao/Data_scratch/Khodursky/sexbias_var/')

```

Load output of gamlss. This table should contain all the information you need about whether a gene
is SDV or not. At this stage use the gamlss output with PC covariates.

```{r}
gamlss_out <- read_csv('gamlss_sum/gamlss_out_PC.csv')

```

Load functions required for gene set analysis

```{r}
source('scripts/geneset_funcs.R')
```
Load biomart information. We need this to convert between ensembl IDs (used for gamlss) 
to entrez IDs which are needed for msigdb data. Some of the entrez IDs apparently dont exist 
for certain ensembl IDs so remove them (na.omit)

```{r}
#ensembl <- useMart(biomart="ensembl",
#                   dataset="hsapiens_gene_ensembl")
# these are all the genes tested by gamlss
#all_genes <- getBM(attributes=c('ensembl_gene_id','entrezgene_id'),
#                    filters = 'ensembl_gene_id', 
#                    values = unique(gamlss_out$gene_id), 
#                   mart = ensembl) %>% na.omit()
all_genes <- read.csv('ensembl_entrez_map.csv', stringsAsFactors = F)  
colnames(all_genes) <- c("gene_id","entrez_id")
#write.csv(all_genes, 'ensembl_entrez_map_new.csv', quote = F, row.names = F)
all_res <- merge(all_genes, gamlss_out, by = "gene_id") %>% distinct()
```

Identify the lists of tissues on which we will perform the gene set analysis.
For instance only consider tissues which have at least 20 male and/or female
high variability genes

```{r}

# identify tissues with at least 20 SDV genes in each class
tissues_20_male <- gamlss_out %>% get_topNtis(kind = 'male high var.', thresh = 20)
tissues_20_female <- gamlss_out %>% get_topNtis(kind = 'female high var.', thresh = 20)



```

split gamlss into list to 
```{r}

tissues <- unique(c(tissues_20_female, tissues_20_male))

gamlss_list <- map(tissues, split_gamlss, all_genes = all_genes, gamlss_out = gamlss_out)
names(gamlss_list) <- tissues
```

Load the hallmark database from msigdb as a dataframe. Excluding gene sets that are highly
tissue/development specific form the meta analysis.
```{r}
m2g_h <- get_m2g('H')
# filter genesets that are tissue specific
to_remove <- c("HALLMARK_PANCREAS_BETA_CELLS",
               "HALLMARK_ADIPOGENESIS",
               "HALLMARK_MYOGENESIS",
               "HALLMARK_SPERMATOGENESIS",
               "HALLMARK_BILE_ACID_METABOLISM",
               "HALLMARK_HEME_METABOLISM",
               "HALLMARK_XENOBIOTIC_METABOLISM")
m2g_h <- m2g_h %>% filter(! (gs_name %in% to_remove))

```

Load permutation results. These were generated by the script 'geneset_hpc.R'. In short 
for each geneset we sum the log p-values across the tested tissues (with at least 20 male or female SDV genes).
The p-values for each geneset-tissue combo were calculated using a hypergeometric test.
This is our test statistic. We then perform 10,000 permutations for each geneset-tissue combo
by permuting the gene labels (male high var, female high var, and SDV). We then compare the empirical null
to the observed value of our test statistic. This is an empirical variation on fisher's method

```{r}
empp_female_hall <- read.csv('geneset_res/empp_hallmark_female_090622.csv', stringsAsFactors = FALSE)
empp_male_hall <- read.csv('geneset_res/empp_hallmark_male_090622.csv', stringsAsFactors = FALSE)
```



Plot significant hallmark gene sets in both sexes after permutation

```{r}
# the number of permutations
n <- 10000
metap_hall <- rbind(empp_female_hall, empp_male_hall) %>% mutate(collect = "Hallmark")
plot_bar_hall <- plot_bar(metap_hall, fdr = 0.05, n)
plot_bar_hall
#ggsave('plots_PC/hallmark_prefilt_bar_10k.pdf', plot_bar_hall, height = 8, width = 6)
```




Perform enrichment meta-analysis for transcription factor targets. First load database
from msigDB. Then proceed in the same exact way as above

```{r}
m2g_reg <- get_m2g('C3','TFT:GTRD')

```


Load the regulatory target permutation results. Identical to the analysis described above with the exception
that the msigdb GTRD database was used


```{r}
empp_female_reg <- read.csv('geneset_res/empp_reg_female_090622.csv', stringsAsFactors = FALSE)
empp_male_reg <- read.csv('geneset_res/empp_reg_male_090622.csv', stringsAsFactors = FALSE)
```





Plot the resulting plots for regulatory targets

```{r}

metap_reg <- rbind(empp_female_reg, empp_male_reg) %>% mutate(collect = "Regulation")
plot_bar_reg<- plot_bar(metap_reg, fdr = 0.05, n, jama=TRUE)
plot_combined <- plot_grid(plot_bar_hall, plot_bar_reg, 
                           labels = c('A', 'C'), ncol = 1, 
                           align = 'v', rel_heights = c(1, 0.18))
plot_combined
plot_bar_reg
```

Plot single tissue dotplots

```{r}
m2g_h <- get_m2g('H')
breast_m <- enrich_tissue("Breast-MammaryTissue",
                           bias_sex = 'male high var.',
                           gamlss_list = gamlss_list,
                           all_sets = m2g_h, 
                           meta = FALSE) %>% 
  mutate(padj = p.adjust(p, method = "BH"),
         tissue = "Breast: male high variability")
plot_singt <- plot_enrichment_tis(breast_m, fdr = 0.05)
plot_singt

```

function for turning p values into text
```{r}
p2text <- function(p) {
 if (p <= 0.05) {
   p <- format(signif(p, 2), scientific = TRUE)
  return(paste0('p=',p))
  
   } else {
    return('')
  }

}

```

single cell analysis of breast epithelial tissue. Summary of results
```{r}

cell_tau <- read.csv('cell_tau.csv', stringsAsFactors = FALSE)

breast <- gamlss_out %>% filter(tissue == 'Breast-MammaryTissue')

cell_tau <- merge(cell_tau, breast, by = "gene_id")

# make plots
ptau <- wilcox.test(cell_tau$tau[cell_tau$meta_bias == 'SDV'], 
                    cell_tau$tau[cell_tau$meta_bias != 'SDV'])$p.value %>%
  p2text()

plot_tau <- ggplot(cell_tau, aes(x=tau, color = meta_bias)) +
  stat_ecdf(geom = "step" , size = 1.2) + 
  theme_classic() + 
  scale_color_jama() + 
  ylab('Cumulative frequency') + 
  xlab(expression(tau)) +
  theme(legend.title = element_blank()) + 
  annotate("text", x = 0.5, y = 0.5, label = ptau)
plot_tau


```



Graves' disease analysis

```{r}
thyroid <- gamlss_out %>% filter(tissue == "Thyroid")
# get female-biased SDV
sig_thyroid <- gamlss_out %>% filter(tissue == "Thyroid",
                                     meta_bias == 'SDV' ) %>% filter(coef.male.sig<0)
graves <- read_tsv('gwas-association-downloaded_2022-08-04-EFO_0004237.tsv') %>% filter(`DISEASE/TRAIT` == 'Graves\' disease')

# gene ids are in a weird format
unsplit_geneids <- function(combined_ids){
  # remove spaces
  combined_ids <- str_replace_all(combined_ids, ' ', '')
  return(unlist(str_split(combined_ids, ',')))
}


ingenes <- map(graves$SNP_GENE_IDS, unsplit_geneids) %>% unlist() %>% unique()

flanking_up <- graves$UPSTREAM_GENE_ID
flanking_down <- graves$DOWNSTREAM_GENE_ID

graves_genes <- unique(c(ingenes, flanking_up, flanking_down))




```

Plot Venn Diagram

```{r}
conv <- read_csv('mart_export_HGNC_ENSG_080122.txt')

conv_hgnc <- function(gene_id, conv){
  return(conv$`HGNC symbol`[conv$`Gene stable ID` == gene_id][1])
}
overlapped <- graves_genes[graves_genes %in% sig_thyroid$gene_id]
annotation <- map(overlapped, conv_hgnc, conv=conv) %>% unlist() %>% paste(collapse ='\n')

num_nongraves <- dim(thyroid)[1] - sum(graves_genes %in% thyroid$gene_id)
# hypergeometric test
p <- 1-phyper(length(overlapped) - 1, 
         sum(graves_genes %in% thyroid$gene_id), 
         num_nongraves, 
         dim(sig_thyroid)[1])



thyroid_sdv <- sig_thyroid$gene_id
graves_expressed <- graves_genes[graves_genes %in% thyroid$gene_id]

plot_graves <- ggvenn(list(`SDV\ngenes`= thyroid_sdv, `Graves'\ngenes `= graves_expressed), show_percentage = FALSE) +  
  annotate("text", x=0, y=-1.4, label=annotation, size=3) +  
  annotate("text", x=1.3, y=-1.3, label=p2text(p), size=3) 


plot_graves

```



Combine all of the plots into one
```{r}
plot_right <- plot_grid(plot_singt, plot_tau, NULL , ncol = 1, labels = c('B', 'D', 'E'), axis = "l",
                        align = "v", rel_heights = c(1, 1, 1))

plot_final <- plot_grid(plot_combined, plot_right, ncol = 2, rel_widths = c(0.9, 1))
ggsave('plots/biological_prop_plot.pdf', plot_final, height = 10, width = 14)
ggsave('plots/graves_venn.pdf', plot_graves, height = 4, width = 6)

```


